tmt/steps/__init__.py

```diff
@@ -1307,3 +1314,305 @@ def after_test(
+            on.
+        """
+
+        assert isinstance(queued_phase.phase, Plugin)  # narrow type
+
+        multiple_guests = len(queued_phase.guests) > 1
+
+        environment = cls.prepare_environment(queued_phase.guests)
+        loggers = cls.prepare_loggers(queued_phase.phase._logger, queued_phase.guests)
+
+        with ThreadPoolExecutor(max_workers=len(queued_phase.guests)) as executor:
+            futures: Dict[Future[None], Tuple[Guest, tmt.log.Logger]] = {}
+
+            for guest in queued_phase.guests:
+                # Swap guest's logger for the one we prepared, with labels
+                # and stuff.
+                #
+                # We can't do the same for phases - phase is shared among
+                # guests, its `self.$loggingmethod()` calls need to be
+                # fixed to use a logger we pass to it through the executor.
+                #
+                # Possibly, the same thing should happen to guest methods as
+                # well, then the phase would pass the given logger to guest
+                # methods when it calls them, propagating the single logger we
+                # prepared...
+                old_logger, new_logger = guest._logger, loggers[guest.name]
+                guest.inject_logger(new_logger)
+
+                if multiple_guests:
+                    new_logger.info('queued', color='cyan')
+
+                # Submit each phase/guest combination (save the guest & logger
+                # for later)...
+                futures[
+                    executor.submit(
+                        queued_phase.phase.go,
+                        guest=guest,
+                        environment=cls.prepare_guest_environment(environment, guest),
+                        logger=new_logger)
+                    ] = (guest, new_logger)
+
+            # ... and then sit and wait as they get delivered to us as they
+            # finish. Unpack the guest and logger, so we could preserve logging
+            # and prepare the right outcome package.
+            for future in as_completed(futures):
+                guest, new_logger = futures[future]
+
+                if multiple_guests:
+                    new_logger.info('finished', color='cyan')
+
+                # `Future.result()` will either 1. reraise an exception the
+                # callable raised, if any, or 2. return whatever the callable
+                # returned - which is `None` in our case, therefore we can
+                # ignore the return value.
+                try:
+                    future.result()
+
+                except Exception as exc:
+                    yield PhaseOutcome(queued_phase, new_logger, guest=guest, exc=exc)
+
+                else:
+                    yield PhaseOutcome(queued_phase, new_logger, guest=guest)
+
+                # Don't forget to restore the original logger.
+                guest.inject_logger(old_logger)
```

Comment:
This sets the same 'old_logger' to all guests (from the last guest processed in `for guest in queued_phase.guests`)

Reply:
Good catch, fixed in 0ba09f0d.